### ISSUE TITLE
Convert ES6 Constructor to Ember `init` hook

### ISSUE THREAD

**Issue by alexlafroscia** [alexlafroscia/ember-mixinify-class#2]
# The Issue

Currently, when porting the properties of the ES6 class over to the Ember mixin, I [skip over the original constructor](https://github.com/alexlafroscia/ember-mixinify-class/blob/deac5d30994891444568199884556b9274d7f7ca/addon/index.js#L7-L9) because of the check that Babel does to ensure that the ES6 class is not being invoked as a regular function:

``` javascript
// ES6 Class
class MyClass {

}

// ES5-compiled version of the above class
function _classCallCheck(instance, Constructor) {
  if (!(instance instanceof Constructor)) {
    throw new TypeError("Cannot call a class as a function");
  }
}

var MyClass = function MyClass() {
  _classCallCheck(this, MyClass);
};
```

If I port the ES6 constructor directly over to the `init` property of the mixin, the check fails because the Ember object being created is no longer an instance of the Babel class.  While a user of the addon could just define an `init` method on the ES6 class, I think it would be a nice addition to do the work in the addon to port the constructor from one style to the other.
# Attempted Solutions

I've tried two different approaches for this, neither of which really worked.
## Dynamically Create a New Function

The first approach I took was to try removing the class check, so that it can not fail.  I `String`-ified the constructor function, REGEX replace the `_classCallCheck`, and then use the `new Function` constructor to create a new version of the function without the check (I also REGEX the arguments out of the original function and use them in the `new Function` constructor).  While this does work, it also destroys the original closure that the function has access to, so something like the following would fail:

``` javascript
import Ember from 'ember';

const { set } = Ember;

class MyClass {
  constructor(name) {
    set(this, 'name', name); // `set` will be undefined after the above processing takes place
  }
}
```

I don't find this approach acceptable, since a user of the addon would not be able to rely on things that are core parts of the JS language.  If the function they want is attached to class itself and accessible as `this._____`, then they would be fine, but not being able to access anything else doesn't feel like a good solution. 
## Modify the Prototype Chain

The second approach I tried was to try getting the class check to pass instead of removing it entirely.  I got this working too, but it required that I wrap the `init` function, walk down the `__proto__` chain until I reached the end (either `undefined` or `Object`) and then replace that with an instance of the original ES6 class.  However, I know that operating on `__proto__` in that was is not recommended for performance reasons, so I don't find it to be an acceptance solution either.

If anyone has any suggestions on a way to pull this off, I would love some feedback on the problem!  I've put some of the approaches I've tried so far in [this gist](https://gist.github.com/alexlafroscia/ee63ea6aa7fe680ec051).

**Comment by joeyjiron06** [comment]
hey any updates on this issue? i am very interested in this addon, although the constructor problem is a huge issue. just wondering what kind of performance issues there are with operating on `__proto__`?

**Comment by alexlafroscia** [comment]
I don't think this will ever be closed; in fact, it was developed for `ember-ajax` but is no longer in use over there.

It's not so much that I specifically saw performance issue with `__proto__`, but I did a whole bunch of research on it and saw over and over that you shouldn't mess with it.

If you have control over the ES6 class, the solution I found was to make the constructor immediately call an `init` method that actually does whatever setup is required.  However, there's no good way to handle arguments to the constructor that I've found.
